cloud/openstack/os_keystone_domain.py

```diff
@@ -60,60 +59,86 @@
 - os_keystone_domain: name=demo description="Default User"
 '''
 
+def _system_state_change(module, domain, new_domain):
+    state = module.params['state']
+    if state == 'present' and \
+       (domain is None or not compare_domain(domain, new_domain)):
+        return True
+    if state == 'absent' and domain is not None:
+        return True
+    return False
+
+def compare_domain(domain_a, domain_b):
+    return domain_a['name'] == domain_b['name'] and \
+           domain_a['enabled'] == domain_b['enabled'] and \
+           domain_a['description'] == domain_b['description']
 
 def main():
 
     argument_spec = openstack_full_argument_spec(
-    argument_spec = dict(
         name=dict(required=True),
         description=dict(default=None),
         enabled=dict(default=True, type='bool'),
         state=dict(default='present', choices=['absent', 'present']),
-    ))
+    )
 
     module_kwargs = openstack_module_kwargs()
-    module = AnsibleModule(argument_spec, **module_kwargs)
+    module = AnsibleModule(argument_spec,
+                           supports_check_mode=True,
+                           **module_kwargs)
 
     if not HAS_SHADE:
         module.fail_json(msg='shade is required for this module')
 
-    name = module.params.pop('name')
-    description = module.params.pop('description')
-    enabled = module.params.pop('enabled')
-    state = module.params.pop('state')
+    name = module.params['name']
+    description = module.params['description']
+    enabled = module.params['enabled']
+    state = module.params['state']
 
     try:
         cloud = shade.operator_cloud(**module.params)
 
         domain = cloud.get_identity_domain(name=name)
+        domain_kwargs = {
+            'name': name,
+            'enabled': enabled,
+            'description': description
+        }
+
+        if module.check_mode:
+            module.exit_json(changed=_system_state_check(module,
+                                                         service,
+                                                         domain_kwargs))
 
         if state == 'present':
             if domain is None:
-                domain = cloud.create_domain(
-                    name=name, description=description, enabled=enabled)
-                changed = True
+                domain = cloud.create_domain(**domain_kwargs)
+                module.exit_json(changed=True,
+                                 result='created',
+                                 id=domain['id'])
+            elif not compare_domains(domain, domain_kwargs):
+                # domains did not match
+                cloud.update_domain(domain_kwargs)
+                module.exit_json(changed=True,
+                                 result='updated',
```

Comment:
Ah thanks. I really don't like having result in there.

Though I can see it being useful because there are some returns that you can't determine what happened with it. That being said it is not normal ansible an I am happy to remove it